### PR TITLE
Add NFE counting for odeint

### PR DESCRIPTION
**Pull request by jacobjinkelly**
Counting NFE on the forward pass is relatively straightforward. I modified`odeint`to return a tuple of `(NFE, ys)` instead of `ys`. To my understanding, switching the order to `(ys, NFE)` would allow using `has_aux=True` when calling `jax.grad`.

Counting NFE on the backward pass is more tricky. The approach I took was to lift a dummy variable to the call signature of `_odeint_wrapper` which gives me a spot in the output of the `vjp` function to put NFE for the backward pass. 

Is there a better solution? A similar problem comes up in counting the number of steps used by a fixed-point solver both on its "forward" and "backward" passes (outlined in [Custom derivative rules for JAX-transformable Python functions](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html)).

Notes:

- NFE = Number of Function Evaluations of solver
- The output of `_odeint_wrapper` is now a tuple, which needs to be cast into a JAX-type. Is there a better way to accomplish this than ravelling and unravelling with `ravel_pytree`

### CODE REVIEW

**Comment by shoyer** [comment]
Very nice, we definitely need support for this! I want the same thing for iterative linear solvers like conjugate gradients.

If I understand correctly, you are putting the evaluation counter in the cotangents? Is is possible to get this information out if you're using `odeint` as part of a larger function on which you're calling `grad`? It feels like we might need a more comprehensive solution here. Right now my guess is that this wouldn't work.

I also wonder if we ever might want to return other metadata from the solver, like the chosen timesteps. Some of these fields could have a float dtype, but they won't have valid derivatives defined -- at least that could be computed without differentiating through the solver itself, which the entire point of these custom gradient rules is to avoid. It seems like we really should have some way of indicating such terms are "non-differentiable".

**Comment by jacobjinkelly** [comment]
Yes, I'm putting the evaluation counter in the cotangents. To get the information out of the grad of a larger function using `odeint` it must have a dummy argument for that information (`init_nfe` in my case), which is then threaded through to the call of `odeint`. If you take grad wrt that dummy argument you get the auxiliary information. Essentially I'm putting auxiliary information in place of a derivative wrt the dummy argument.

I totally agree that this solution won't work in general, and what we really need is a general option for a vjp to return auxiliary information which is separate from the auxiliary information returns with the primals on the forward pass.

**Comment by samuela** [comment]
This PR makes me wonder: What is the idiomatic approach to handling auxiliary data/results in JAX, both in terms of primal functions and their vjp's? 

I'm in roughly the same boat as @jacobjinkelly... I also need to get auxiliary results out of `odeint`, but it's not clear what the right approach to doing so is.

**Comment by samuela** [comment]
Getting auxiliary info out of vjps seems like the messier of the two problems. Like how would it work with `value_and_grad`? And what about higher-order gradients etc etc...

**Comment by mattjj** [comment]
I am wondering the same, @samuela. Thoughts welcome, though @jekbradbury has led an internal group brainstorming about this (e.g. including Flax and Haiku folks) and it's very tricky.

**Comment by samuela** [comment]
@mattjj This seems like the right use case for something like algebraic effects: Code can be instrumented by adding effects for the values of interest and then it's up to client code to register effect handlers for the things they care about. This way effects do not get in the way of function return types and users can be selective in the information they want, and how it should be processed. Presumably some people just want to log debug info, others may want to record it, etc.

The challenge I foresee in such a system would be getting this to play nicely with XLA which presents a bit of a challenge. XLA has only limited support for side effects. Custom calls are one option but they'd be incredibly slow if you have to jump back out into Python in the middle of a hot loop. Perhaps there could be some way to trace effects and their handlers? XLA is great for speedy numerical code, but it does makes this all a little tricky.

**Comment by shoyer** [comment]
> The challenge I foresee in such a system would be getting this to play nicely with XLA which presents a bit of a challenge. XLA has only limited support for side effects. Custom calls are one option but they'd be incredibly slow if you have to jump back out into Python in the middle of a hot loop. Perhaps there could be some way to trace effects and their handlers? XLA is great for speedy numerical code, but it does makes this all a little tricky.

I think a tracing based approach that rewrites code so it looks functionally pure from the prospective of XLA here is _probably_ the right approach. This is what @jekbradbury used in https://github.com/google/jax/pull/1803 and I think @sschoenholz has explored some prototypes along the same lines as well.

**Comment by jacobjinkelly** [comment]
As mentioned in #2628 , `experimental.host_callback` should provide a way to count NFE (and get other auxiliary information!) from `odeint`.
